server/src/main/java/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -1148,6 +1151,93 @@ public PrimaryDataStoreInfo updateStoragePool(UpdateStoragePoolCmd cmd) throws I
+            throw new InvalidParameterValueException("Invalid scope " + newScope.toString() + "for Primary storage");
+        }
+
+        StoragePoolVO primaryStorage = _storagePoolDao.findById(id);
+        if (primaryStorage == null) {
+            throw new IllegalArgumentException("Unable to find storage pool with ID: " + id);
+        }
+
+        if (!primaryStorage.getStatus().equals(StoragePoolStatus.Disabled)) {
+            throw new InvalidParameterValueException("Scope of the Primary storage with id "
+                    + primaryStorage.getUuid() +
+                    " cannot be changed, as it is not in the Disabled state");
+        }
+
+        ScopeType currentScope = primaryStorage.getScope();
+        if (currentScope.equals(newScope)) {
+            throw new InvalidParameterValueException("New scope must be different than the current scope");
+        }
+
+        HypervisorType hypervisorType;
+        if (currentScope.equals(ScopeType.CLUSTER)) {
+            /*
+             * For cluster wide primary storage the hypervisor type might not be set.
+             * So, get it from the clusterVO.
+             */
+            Long clusterId = primaryStorage.getClusterId();
+            ClusterVO clusterVO = _clusterDao.findById(clusterId);
+            hypervisorType = clusterVO.getHypervisorType();
+        } else {
+            hypervisorType = primaryStorage.getHypervisor();
+        }
+        if (!zoneWidePoolSupportedHypervisorTypes.contains(hypervisorType)) {
+            throw new InvalidParameterValueException("Primary storage scope change is not supported for hypervisor type " + hypervisorType);
+        }
+
+        String providerName = primaryStorage.getStorageProviderName();
+        DataStoreProvider storeProvider = _dataStoreProviderMgr.getDataStoreProvider(providerName);
+        PrimaryDataStoreLifeCycle lifeCycle = (PrimaryDataStoreLifeCycle) storeProvider.getDataStoreLifeCycle();
+        DataStore primaryStore = _dataStoreMgr.getPrimaryDataStore(id);
+
+        Long zoneId = primaryStorage.getDataCenterId();
+        DataCenterVO zone = _dcDao.findById(zoneId);
+        if (zone == null) {
+            throw new InvalidParameterValueException("Unable to find zone by id " + zoneId);
+        }
+        if (zone.getAllocationState().equals(Grouping.AllocationState.Disabled)) {
+            throw new PermissionDeniedException("Cannot perform this operation, Zone is currently disabled: " + zoneId);
+        }
+
+        if (newScope.equals(ScopeType.ZONE)) {
+            ClusterScope clusterScope = new ClusterScope(primaryStorage.getClusterId(), null, zoneId);
+            lifeCycle.changeStoragePoolScopeToZone(primaryStore, clusterScope, hypervisorType);
+
+        } else {
+            Long clusterId = cmd.getClusterId();
+            if (clusterId == null) {
+                throw new IllegalArgumentException("Unable to find cluster with ID: " + clusterId);
```

Comment:
normally we use CloudRuntimeException or InvalidParameterValueException
`clusterId` is null, so no need to add in the message of the exception

Reply:
Done